src/test/java/com/haxwell/apps/quizki/services/QuestionServiceImplTest.java

```diff
@@ -274,5 +335,71 @@ public void getQuestionByIdTest() {
 		assertThat(outputDTO.getId(), equalTo(1l));
 		
 	}
+	
+	@Test
+	public void getQuestionsTest() {
+		
+		
+		/*
+		 * Stated requirements
+		 * URL: /api/question?page=1&size=10
+		 * URL: /api/question?page=1 (size defaults to 10)
+		 * URL: /api/question?size=10 (page defaults to 1)
+		 * URL: /api/question (page and size are default)
+		 * 
+		 * parameters are passed to getQuestions() as received by the controller with defaults
+		 */
+		
+		int n = 17;		//TODO: refactor the magic numbers in the assertions based on this value
+		questions = getQuestions(n);
+		
+		when(questionRepo.count()).thenReturn((long)n);
+		
+		when(questionRepo.findAll(any(Pageable.class))).thenAnswer(new Answer<Page<Question>>() {
+			 
+            public Page<Question> answer(InvocationOnMock invocation) throws Throwable {
+                
+            	
+            	Pageable pageRequest = invocation.getArgument(0);
+            	
+                return getPageQuestions(questions, pageRequest);
+            }
+        });
+		
+		
+		try {
+			outputDTOs = qsImpl.getQuestions(1, 10);
+		} catch (GetQuestionException e) {
+			fail("getQuestions failed");
+		}
+
+		assertThat(outputDTOs.size(), equalTo(10));
+		assertThat(outputDTOs.get(0).getId(), equalTo(1L));
+		assertThat(outputDTOs.get(9).getId(), equalTo(10L));
+		
+		outputDTOs.clear();
+		
+		try {
+			outputDTOs = qsImpl.getQuestions(3, 10);
```

Comment:
So, actually, 
- since we have multiple cases we're testing here, 
- and the service is not going to keep state, 

that means 
- the service is going to make the call for the data, each time we test a given case. 

SO. You will need to use the `.thenReturn().thenReturn().then....` chain in your mock above. And for each `thenReturn()` you supply the data in that case.

I think this works, and covers the same code and logic, in three separate tests. (?)

Reply:
I'm checking that I get the number of questions I expect and that the index of the first and last question in the list is also expected. It's not three different sets of questions. I think this is right but learning the thenReturn method is cool!

Reply:
ok, pardon the airport confusion in my comment above (was sitting in a food court and maybe not too focused!). I see what you are saying, there should be 3 calls to the repo , one for each of the try statements, or 3 `.thenReturn()` statements executed.

I think that is happening. The statement `outputDTOs = qsImpl.getQuestions(3, 10);` calls the `QuestionServiceImpl::getQuestions` method which in turn calls the repo on line 274
`Page<Question> questions = questionRepo.findAll(pageable);`
which calls the mock (as mentioned in the previous comments on this line)
so this results in the `.thenAnswer(...)` method of the mock being called 3 times. I needed to use the `.thenAnswer()` method  instead of the `.thenReturn()` so I could handle the `Pagable` parameter in the call.

Reply:
I can't see how chaining would be implemented here, I'm not familiar with it.